{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from tqdm import tqdm\n",
    "\n",
    "import tensorflow_probability as tfp\n",
    "from tensorflow.keras import backend as K\n",
    "import pdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DataPreprocess():\n",
    "    def __init__(self, datadir, celltypes, bkdata_path, gene_list_path):\n",
    "        '''\n",
    "        Creates preprocessed instance of input data\n",
    "        scdata should be in matrix.mtx within specified folders along with barcodes.tsv and genes.tsv\n",
    "        bkdata should have sample names as columns and gene names as rows\n",
    "        gene_list should have no row or column names/index\n",
    "        '''\n",
    "        self.datadir = datadir\n",
    "        self.celltypes = celltypes\n",
    "        self.scdata = self.load_scdata(self.datadir, self.celltypes)\n",
    "        self.bkdata = pd.read_csv(bkdata_path, sep='\\t')\n",
    "        # If there is input gene list, filter out genes not in bkdata or scdata\n",
    "        if gene_list_path is not None:\n",
    "            self.genelist = pd.read_csv(gene_list_path, squeeze=True, names=['name'])\n",
    "            self.genelist = self.genelist.drop_duplicates()\n",
    "            self.genelist = self.genelist[self.genelist.isin(self.bkdata.index.drop_duplicates())]\n",
    "        else:\n",
    "            self.genelist = self.bkdata.index.drop_duplicates()\n",
    "        self.genelist = self.genelist[self.genelist.isin(self.scdata.var.index)]\n",
    "        self.genelist = self.genelist.sort_values()\n",
    "        # Filter out genes not in gene list\n",
    "        self.scdata = self.scdata[:,self.scdata.var.index.isin(self.genelist)]\n",
    "        sc.pp.normalize_total(self.scdata, target_sum=1e6) # normalize to sum to 1,000,000\n",
    "        # sc.pp.regress_out(scdata, ['total_counts'], n_jobs=1)\n",
    "        # Transpose, filter out genes not in gene list, then sort column (by gene name)\n",
    "        self.bkdata = self.bkdata.T.loc[:,self.genelist].sort_index(axis=1)\n",
    "    def load_scdata(self, data_directories, cell_types):\n",
    "        # Read and merge 10X Genomics scRNA-seq data\n",
    "        scdata = None\n",
    "        for d, c in zip(tqdm(data_directories), cell_types):\n",
    "            x = sc.read_10x_mtx(d)\n",
    "            x.obs['celltype'] = [c]*len(x.obs.index)\n",
    "            # Change each observation (cell) name to celltype + barcode\n",
    "            x.obs.set_index(pd.Index([c+'_'+rn[:-2] for rn in x.obs.index]), inplace=True)\n",
    "            if scdata is not None:\n",
    "                scdata = ad.concat([scdata, x])\n",
    "            else:\n",
    "                scdata = x\n",
    "        # Filter out cells and genes\n",
    "        sc.pp.filter_cells(scdata, min_genes=200)\n",
    "        sc.pp.filter_genes(scdata, min_cells=1)\n",
    "        # Search for prefix \"MT-\" (mitochondrial genes) and make new column in variable annotations\n",
    "        # Search for prefix \"RPL/RPS\" for ribosomal genes and \"MRPL/MRPS\" for mitochondrial ribosomal genes\n",
    "        scdata.var['mito'] = scdata.var.index.str.match('^MT-')\n",
    "        scdata.var['ribo'] = scdata.var.index.str.startswith(('RPL','RPS'))\n",
    "        scdata.var['mribo'] = scdata.var.index.str.startswith(('MRPL','MRPS'))\n",
    "        # Calculate QC metrics as per McCarthy et al., 2017 (Scater)\n",
    "        sc.pp.calculate_qc_metrics(scdata, qc_vars=['mito','ribo', 'mribo'], inplace=True)\n",
    "        # Plot QC metrics\n",
    "        # sns.jointplot(x='total_counts', y='n_genes_by_counts', height=8, data=scdata.obs,\n",
    "        #     kind='scatter', hue='celltype')\n",
    "        # sns.jointplot(x='total_counts', y='pct_counts_mito', height=8, data=scdata.obs,\n",
    "        #     kind='scatter', hue='celltype')\n",
    "        # sns.jointplot(x='total_counts', y='pct_counts_ribo', height=8, data=scdata.obs,\n",
    "        #     kind='scatter', hue='celltype')\n",
    "        # sns.jointplot(x='total_counts', y='pct_counts_mribo', height=8, data=scdata.obs,\n",
    "        #     kind='scatter', hue='celltype')\n",
    "        # plt.show()\n",
    "        # Filter out cells with >5% of counts from mitochondria and mitoribosome\n",
    "        # scdata = scdata[scdata.obs.pct_counts_ribo > 30, :]\n",
    "        scdata = scdata[scdata.obs.pct_counts_mito < 5, :]\n",
    "        scdata = scdata[scdata.obs.pct_counts_mribo < 1, :]\n",
    "        return scdata\n",
    "    def __call__(self, whichdata):\n",
    "        if whichdata == 'scdata':\n",
    "            out = []\n",
    "            for c in tqdm(self.celltypes):\n",
    "                out.append(self.scdata[self.scdata.obs.celltype==c].to_df().sort_index(axis=1))\n",
    "        elif whichdata == 'bkdata':\n",
    "            out = self.bkdata\n",
    "        elif whichdata == 'genelist':\n",
    "            out = self.genelist\n",
    "        else:\n",
    "            raise ValueError('Choose only one of the following: \"scdata\", \"bkdata\", or \"genelist\"')\n",
    "        return out"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
